### PR TITLE
Ingest from NCBI Virus + NCBI Datasets

### PR DESCRIPTION
**Pull request by joverlee521**
## Description of proposed changes

NCBI Datasets does not include the `strain`, `serotype`, and `segment` in the metadata, so this PR uses the NCBI Virus `vvsearch2` API to pull down the metadata and joins it with the sequences downloaded via NCBI Datasets. 

On this branch, I can run the ingest workflow with 

```
nextstrain build ingest ingest_ncbi --configfile build-configs/ncbi/defaults/config.yaml
```

then successfully run the genome build (~using tmp changes in 4b511fc25620ea1fd6ef24846372ffc124344134~)

```
nextstrain build . --snakefile Snakefile.genome --config local_ingest=True ingest_source=ncbi
```

~Still needs some clean up but at least wanted to get this out for people to try out.~

## Major change

This PR completely changes over the h5n1-cattle-outbreak build to the NCBI data. 
See the latest build at https://nextstrain.org/staging/avian-flu/ncbi/h5n1-cattle-outbreak/genome

## Related issue(s)

Related to #37 

## Checklist

- [x] Checks pass
- [x] Add description.md for h5n1-cattle-outbreak
- [x] phylo include/exclude strain names
- [x] Add uploads to S3 -> tracking in https://github.com/nextstrain/avian-flu/issues/41
- [x] join NCBI w/ Anderson lab data -> tracking in https://github.com/nextstrain/avian-flu/issues/42
- [x] add ingest automation -> tracking in https://github.com/nextstrain/avian-flu/issues/43
- [x] `nextclade run` as part of ingest -> tracking in https://github.com/nextstrain/avian-flu/issues/44

### CODE REVIEW

File: ingest/build-configs/ncbi/rules/curate.smk
Comment (joverlee521) [inline ingest/build-configs/ncbi/rules/curate.smk]: Noted by @jameshadfield on [Slack](https://bedfordlab.slack.com/archives/CD84ELG0N/p1716761204528389?thread_ts=1716605172.691119&cid=CD84ELG0N)

```suggestion
        subset_metadata="ncbi/data/metadata_{segment}.tsv",
```

File: ingest/build-configs/ncbi/rules/curate.smk
Comment (joverlee521) [inline ingest/build-configs/ncbi/rules/curate.smk:77]: Noted by @jameshadfield on [Slack](https://bedfordlab.slack.com/archives/CD84ELG0N/p1716761285536859?thread_ts=1716605172.691119&cid=CD84ELG0N)

This outputs a _lot_ of warning messages when for column names that are the same. Can use the `--force` flag to silence the warnings.

Comment (joverlee521) [reply ingest/build-configs/ncbi/rules/curate.smk:77]: Resolving upstream in https://github.com/nextstrain/ingest/pull/38. 

Will update the vendored scripts here once it's merged.

File: ingest/build-configs/ncbi/rules/curate.smk
Comment (joverlee521) [reply ingest/build-configs/ncbi/rules/curate.smk]: Updated in https://github.com/nextstrain/avian-flu/pull/40/commits/d5db437e33f8150a751303722e75515f194acaa8

File: ingest/build-configs/ncbi/rules/curate.smk
Comment (joverlee521) [reply ingest/build-configs/ncbi/rules/curate.smk:77]: Merged and updated vendored scripts in https://github.com/nextstrain/avian-flu/pull/40/commits/a2b01e7e2b351fa067b8ad694481e10cd872318d.

**Comment by joverlee521** [comment]
The h5n1-cattle-outbreak build is still using the shared [config/description.md](https://github.com/nextstrain/avian-flu/blob/master/config/description.md). I think we should have a separate description.md that points to the public data sources.

**Comment by trvrb** [comment]
Nice work here @joverlee521! I was able to run this locally and everything worked in a clean fashion. The revised exclude list looks appropriate. After revision, the only blocking issue that I see is that `division` isn't coming through for most of the sequences. For example `A/turkey/Missouri/24-005369-001/2024` lists `division` as `USA`. You can see this issue here: https://nextstrain.org/staging/avian-flu/ncbi/h5n1-cattle-outbreak/genome?c=division.

I don't see "Missouri" listed anywhere on accession [PP748223](https://www.ncbi.nlm.nih.gov/nuccore/PP748223). But conversely, `A/Bovine/texas/24-029328-01/2024` lists `division` as `Texas`, while the accession [PP599465](https://www.ncbi.nlm.nih.gov/nuccore/PP748223) lists country as `USA: texas`.

Is there a way to more thoroughly collect division information via NCBI? Or do we need to somewhere scrape this from strain name during ingest?

**Review by trvrb — CHANGES_REQUESTED** [review]
This is excellent @joverlee521. As in the above comment, the only blocking issue I see is the lack of division-level metadata.

**Comment by joverlee521** [comment]
> I don't see "Missouri" listed anywhere on accession [PP748223](https://www.ncbi.nlm.nih.gov/nuccore/PP748223). But conversely, A/Bovine/texas/24-029328-01/2024 lists division as Texas, while the accession [PP599465](https://www.ncbi.nlm.nih.gov/nuccore/PP748223) lists country as USA: texas.

Hmm, I see that `A/Bovine/texas/24-029328-01/2024` has the [correct division in the build](https://nextstrain.org/staging/avian-flu/ncbi/h5n1-cattle-outbreak/genome?c=division&f_genbank_accession=PP599465).

> Is there a way to more thoroughly collect division information via NCBI? Or do we need to somewhere scrape this from strain name during ingest?

This is all the data that's available through the usual metadata from NCBI, I'll just have to scrape the strain name for additional data.

**Comment by trvrb** [comment]
>This is all the data that's available through the usual metadata from NCBI, I'll just have to scrape the strain name for additional data.

This makes sense. I don't see a way around this. Sorry that this is so difficult.

File: README.md
Comment (genehack) [inline README.md]:
```suggestion
Currently this is only set up for the "h5n1-cattle-outbreak" build using locally ingested NCBI data, and the build is restricted to a set of strains where we think there's no reassortment, with outgroups
```

File: Snakefile.genome
Comment (genehack) [inline Snakefile.genome:178]: is it worth a comment here about why this strain for the root? feels kinda magic-number-y

File: ingest/build-configs/ncbi/bin/ncbi-virus-url
Comment (genehack) [inline ingest/build-configs/ncbi/bin/ncbi-virus-url]:
```suggestion
def parse_args() -> argparse.Namespace:
```

File: ingest/build-configs/ncbi/bin/ncbi-virus-url
Comment (genehack) [inline ingest/build-configs/ncbi/bin/ncbi-virus-url]:
```suggestion
def build_query_url(ncbi_taxon_id: str,
                    filters: Optional[List[str]]=None,
                    fields: Optional[List[str]]=None) -> None:
```

File: ingest/build-configs/ncbi/bin/parse-metadata-from-strain
Comment (genehack) [inline ingest/build-configs/ncbi/bin/parse-metadata-from-strain]:
```suggestion
def update_record(record: Dict[str, str], strain_metadata: Dict[str, str]) -> None:
```

File: ingest/build-configs/ncbi/bin/transform-host
Comment (genehack) [inline ingest/build-configs/ncbi/bin/transform-host]:
```suggestion
def parse_args() -> argparse.Namespace:
```

**Comment by joverlee521** [comment]
[Division](https://nextstrain.org/staging/avian-flu/ncbi/h5n1-cattle-outbreak/genome?c=division) is more filled out with metadata from the strain name using https://github.com/nextstrain/avian-flu/pull/40/commits/19bc00eed3ac64178f3844628befd1816e27dc87: 

Screenshot 2024-05-29 at 3 19 13 PM

---

I'll plan to merge this tomorrow if there are no other comments.

File: ingest/build-configs/ncbi/bin/transform-segment
Comment (genehack) [inline ingest/build-configs/ncbi/bin/transform-segment]:
```suggestion
def parse_args() -> argparse.Namespace:
```

File: ingest/build-configs/ncbi/rules/curate.smk
Comment (genehack) [inline ingest/build-configs/ncbi/rules/curate.smk]: why the `>>` instead of `>` ?

**Review by genehack — APPROVED** [review]
LGTM; contains several very nitpicky nits picked by a nitwit.

File: ingest/build-configs/ncbi/defaults/annotations.tsv
Comment (trvrb) [inline ingest/build-configs/ncbi/defaults/annotations.tsv]: Entries in `annotations.tsv` are not currently being applied to the fetched and parsed metadata. If I inspect `ingest/ncbi/results/metadata.tsv` I get division=USA for `A/environment/USA/CO-UW-9084466/2024`.

**Review by trvrb — CHANGES_REQUESTED** [review]
Your script to parse location from strain name was very clever!

This is basically good to go for me. There was just one thing I caught where annotations don't seem to be being applied.

File: Snakefile.genome
Comment (trvrb) [reply Snakefile.genome:178]: It's the closest outgroup.

File: ingest/build-configs/ncbi/defaults/annotations.tsv
Comment (joverlee521) [reply ingest/build-configs/ncbi/defaults/annotations.tsv]: Oh, good catch. Will debug 🤔

File: ingest/build-configs/ncbi/rules/curate.smk
Comment (joverlee521) [reply ingest/build-configs/ncbi/rules/curate.smk]: Good question. This was copied from [pathogen-repo-guide](https://github.com/nextstrain/pathogen-repo-guide/blob/80cc1733b47f173772fe45005b82a69c6f8cb6c9/ingest/rules/curate.smk#L39), which was based on [mpox/ingest](https://github.com/nextstrain/mpox/blob/7f5adc3cab0c3034e455882581cc081f26f5ebbe/ingest/rules/curate.smk#L35) and leads back to [my own commit from ~2 years ago](https://github.com/nextstrain/mpox/commit/67aa5f2cbb08b9ef0d3ec75b78518786d59f7cbe). 

What was I thinking 🙃 My best guess is this is left over from when I had a rule to just append the `local_geolocation_rules` to the existing `general_geolocation_rules` without creating the additional `all_geolocation_rules` file.

File: ingest/build-configs/ncbi/defaults/annotations.tsv
Comment (joverlee521) [reply ingest/build-configs/ncbi/defaults/annotations.tsv]: Fixed in https://github.com/nextstrain/avian-flu/pull/40/commits/f04d275d87472ff603b7cc7a646407a193bd541d.

I forgot that the format of the annotations.tsv only allows 2 tabs, so final comments need to be separated by a space.

File: Snakefile.genome
Comment (joverlee521) [reply Snakefile.genome:178]: Added comment in https://github.com/nextstrain/avian-flu/pull/40/commits/b477716785e31c092ae75e6d9bde30ae7ee2d763

File: ingest/build-configs/ncbi/rules/curate.smk
Comment (joverlee521) [reply ingest/build-configs/ncbi/rules/curate.smk]: Updated in https://github.com/nextstrain/avian-flu/pull/40/commits/096ce904b8599227376294d4a691ad4ecde8ef61

File: README.md
Comment (joverlee521) [reply README.md]: Included in https://github.com/nextstrain/avian-flu/pull/40/commits/eb5e068ef2d4260b01c7ac49fa805b539fb05e83

File: ingest/build-configs/ncbi/bin/ncbi-virus-url
Comment (joverlee521) [reply ingest/build-configs/ncbi/bin/ncbi-virus-url]: Included in https://github.com/nextstrain/avian-flu/pull/40/commits/eb5e068ef2d4260b01c7ac49fa805b539fb05e83

File: ingest/build-configs/ncbi/bin/ncbi-virus-url
Comment (joverlee521) [reply ingest/build-configs/ncbi/bin/ncbi-virus-url]: Included in https://github.com/nextstrain/avian-flu/pull/40/commits/eb5e068ef2d4260b01c7ac49fa805b539fb05e83

File: ingest/build-configs/ncbi/bin/parse-metadata-from-strain
Comment (joverlee521) [reply ingest/build-configs/ncbi/bin/parse-metadata-from-strain]: Included in https://github.com/nextstrain/avian-flu/pull/40/commits/eb5e068ef2d4260b01c7ac49fa805b539fb05e83

File: ingest/build-configs/ncbi/bin/transform-host
Comment (joverlee521) [reply ingest/build-configs/ncbi/bin/transform-host]: Included in https://github.com/nextstrain/avian-flu/pull/40/commits/eb5e068ef2d4260b01c7ac49fa805b539fb05e83

File: ingest/build-configs/ncbi/bin/transform-segment
Comment (joverlee521) [reply ingest/build-configs/ncbi/bin/transform-segment]: Included in https://github.com/nextstrain/avian-flu/pull/40/commits/eb5e068ef2d4260b01c7ac49fa805b539fb05e83

File: ingest/build-configs/ncbi/defaults/annotations.tsv
Comment (trvrb) [reply ingest/build-configs/ncbi/defaults/annotations.tsv]: Ah! Nice catch. Thanks for fixing.

**Pull request merged**
